### PR TITLE
fix(ui): don't use `Buffer` for FNV hash

### LINKED ISSUE #11764: [master branch only] UI crashes when clicking node with an error message: Buffer is not defined
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

THIS DOES NOT RELATES TO RELEASED IMAGES either v3.5.0-rc nor v3.10.0.
Might be local development only?

* Summary
  * UI crashes when we click a node `hello1`.
  * Curiously, `step-mnss7` is ok.
  * image
  * image
* Procedure
  * checkout master branch
  * `make start UI=true`
* My Env
  * WSL
  * k3d

### Version

master

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: steps-
spec:
  entrypoint: hello-hello-hello

  # This spec contains two templates: hello-hello-hello and whalesay
  templates:
  - name: hello-hello-hello
    # Instead of just running a container
    # This template has a sequence of steps
    steps:
    - - name: hello1            # hello1 is run before the following steps
        template: whalesay
        arguments:
          parameters:
          - name: message
            value: "hello1"
    - - name: hello2a           # double dash => run after previous step
        template: whalesay
        arguments:
          parameters:
          - name: message
            value: "hello2a"
      - name: hello2b           # single dash => run in parallel with previous step
        template: whalesay
        arguments:
          parameters:
          - name: message
            value: "hello2b"

  # This is the same template as from the previous example
  - name: whalesay
    inputs:
      parameters:
      - name: message
    container:
      image: docker/whalesay
      command: [cowsay]
      args: ["{{inputs.parameters.message}}"]
```
```


### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by agilgur5**
Fixes #11764

### Motivation

- `Buffer` requires a polyfill in client-side JS
  - #11628 upgraded to Webpack v5, which [no longer auto adds Node polyfills](https://github.com/webpack/changelog-v5#automatic-nodejs-polyfills-removed)
    - which ended up breaking 2 year old code from #6925
  - this seems to be the only one we use (as far as I can tell) -- and this code quite literally copies server-side Go code, so makes sense that it may use server-side libraries
    - add comments in the code about where this is copied from so that it is easier to track (I had to go through the `blame` + commit history)

### Modifications

- use `charCodeAt` to extract an int from a character of a string instead of converting the string to a Buffer
  - was inspired by looking at some JS FNV implementations and stumbled upon some lines like this: https://github.com/schwarzkopfb/fnv1a/blob/2c5ee2889e42a6bd38c64a29b227eaf9e048e40d/index.js#L22

- also re-enable the `tslint` rule at the end of the code

### Verification

- unit tests for FNV hashing still pass, so think the implementation is equivalent and we're good!

### Misc Notes

I decided not to replace the FNV hash implementation with a library as basically [all](https://www.npmjs.com/search?q=fnv1a) of them are unmaintained and this change only ended up being a few LoC. Some of the libraries are also [quite large](https://github.com/tjwebb/fnv-plus/blob/v1.3.0/index.js) (implementing many different algorithms and are not tree-shakeable), or have [build issues](https://github.com/schwarzkopfb/fnv1a/issues/2), or [use BigInt](https://github.com/sindresorhus/fnv1a/blob/ddbe4f242c6d9a86258400dcf6a1dabae156a3e1/index.js#L5) (which is not supported in older browsers... although those are all EoL and I wouldn't personally support such old browsers, but still something to consider).

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
@toyamagu-2021 Can you help test this change?

File: ui/src/app/shared/pod-name.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ export const k8sNamingHashLength = 10;
 // getPodName returns a deterministic pod name
 // In case templateName is not defined or that version is explicitly set to  POD_NAME_V1, it will return the nodeID (v1)
 // In other cases it will return a combination of workflow name, template name, and a hash (v2)
+// note: this is inteded to be equivalent to the server-side Go code in workflow/util/pod_name.go
```
Comment (agilgur5) [inline ui/src/app/shared/pod-name.ts:12]:
```suggestion
// note: this is intended to be equivalent to the server-side Go code in workflow/util/pod_name.go
```

typo

**Review by toyamagu-2021 — APPROVED** [review]
Thanks! Works fine.

But I wonder why we should calculate podName by nix hash in frontend?
We may derive podName from nodeId...

`nodeId`=`steps-dflj9-1988631110`
`podName`=`steps-dflj9-whalesay-1988631110`

**Review by terrytangyuan — APPROVED** [review]

**Comment by agilgur5** [comment]
> But I wonder why we should calculate podName by nix hash in frontend? We may derive podName from nodeId...
> 
> `nodeId`=`steps-dflj9-1988631110` `podName`=`steps-dflj9-whalesay-1988631110`

[the logic](https://github.com/agilgur5/argo-workflows/blob/38a09001bebcdece74b1a51f0feedef7ec5fc7ff/ui/src/app/shared/pod-name.ts#L18), as far as I can tell, is that when the name is too long it needs to be shortened, so a hash is used. The back-end PR is #6712 which also has a few back-links to other issues and PRs. Could also ask JP if he remembers

I've definitely had some long names before and it's common to do some truncation or hashing when names are too long for k8s's limit. So it makes sense to me.

**Comment by toyamagu-2021** [comment]
Thanks for clarifying!
Now also makes sense to me.

**Pull request merged**